tests/unit/test-class-wc-payments-token-service.php

```diff
@@ -799,4 +807,298 @@ private function generate_link_token( $stripe_id, $wp_id = 0 ) {
+		$new_customer_id        = 'cus_67890';
+		$gateway_id             = 'woocommerce_payments';
+		$cached_payment_methods = [
+			$this->generate_card_pm_response( 'pm_cached1' ),
+		];
+		$cached_data            = [
+			'customer_id'                    => $customer_id,
+			'payment_methods_' . $gateway_id => $cached_payment_methods,
+		];
+
+		// Add cached data with different customer ID.
+		update_user_meta( $user_id, WC_Payments_Token_Service::CACHED_PAYMENT_METHODS_META_KEY, $cached_data );
+
+		$new_payment_methods = [
+			$this->generate_card_pm_response( 'pm_new1' ),
+			$this->generate_card_pm_response( 'pm_new2' ),
+		];
+
+		// Mock gateway to return enabled payment methods.
+		$mock_gateway = $this->createMock( WC_Payment_Gateway_WCPay::class );
+		$mock_gateway->method( 'get_upe_enabled_payment_method_ids' )->willReturn( [ Payment_Method::CARD ] );
+		WC_Payments::set_gateway( $mock_gateway );
+
+		// Mock customer service to return new payment methods.
+		$this->mock_customer_service
+			->expects( $this->once() )
+			->method( 'get_payment_methods_for_customer' )
+			->with( $new_customer_id, Payment_Method::CARD )
+			->willReturn( $new_payment_methods );
+
+		$result = $this->call_sut_method( 'get_payment_methods_from_stripe', $user_id, $new_customer_id, $gateway_id );
+
+		// Verify new payment methods are returned.
+		$this->assertEquals( $new_payment_methods, $result );
+
+		// Verify cache is updated with new customer ID and payment methods.
+		$updated_cache = get_user_meta( $user_id, WC_Payments_Token_Service::CACHED_PAYMENT_METHODS_META_KEY, true );
+		$this->assertEquals( $new_customer_id, $updated_cache['customer_id'] );
+		$this->assertEquals( $new_payment_methods, $updated_cache[ 'payment_methods_' . $gateway_id ] );
+	}
+
+	/**
+	 * Test get_payment_methods_from_stripe method with no cached data.
+	 */
+	public function test_get_payment_methods_from_stripe_with_no_cached_data() {
+		$user_id         = 1;
+		$customer_id     = 'cus_12345';
+		$gateway_id      = 'woocommerce_payments';
+		$payment_methods = [
+			$this->generate_card_pm_response( 'pm_new1' ),
+			$this->generate_link_pm_response( 'pm_new2' ),
+		];
+
+		// Mock gateway to return enabled payment methods.
+		$mock_gateway = $this->createMock( WC_Payment_Gateway_WCPay::class );
+		$mock_gateway->method( 'get_upe_enabled_payment_method_ids' )->willReturn( [ Payment_Method::CARD, Payment_Method::LINK ] );
+		WC_Payments::set_gateway( $mock_gateway );
+
+		// Mock customer service to return payment methods.
+		$this->mock_customer_service
+			->expects( $this->exactly( 2 ) )
+			->method( 'get_payment_methods_for_customer' )
+			->withConsecutive(
+				[ $customer_id, Payment_Method::CARD ],
+				[ $customer_id, Payment_Method::LINK ]
+			)
+			->willReturnOnConsecutiveCalls(
+				[ $payment_methods[0] ],
+				[ $payment_methods[1] ]
+			);
+
+		$result = $this->call_sut_method( 'get_payment_methods_for_customer', $user_id, $customer_id, $gateway_id );
```

Comment:
Why are we calling a method that was just mocked?

Reply:
I changed the method name in https://github.com/Automattic/woocommerce-payments/pull/11092/commits/6eb011e0e031a520ce66b2bdd31749ee19999367, it was always supposed to be `get_payment_methods_from_stripe`. However, even before that, I was calling the `get_payment_methods_for_customer` of the token service, while mocking the method of the customer service. Same name, different methods.